api/v1alpha1/extension_types.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Although this looks like a bool, there might
+type ExtensionPaused string
+
+const (
+	// Pause resolution of this Extension
+	ExtensionPausedTrue ExtensionPaused = "True"
+	// Peform resolution of this Extension
+	ExtensionPausedFalse ExtensionPaused = "False"
+)
+
+type ExtensionSourcePackage struct {
+	//+kubebuilder:validation:MaxLength:=48
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+(-[a-z0-9]+)*$
+	Name string `json:"name"`
+
+	//+kubebuilder:validation:MaxLength:=64
+	//+kubebuilder:validation:Pattern=`^(\s*(=||!=|>|<|>=|=>|<=|=<|~|~>|\^)\s*(v?(0|[1-9]\d*|[x|X|\*])(\.(0|[1-9]\d*|x|X|\*]))?(\.(0|[1-9]\d*|x|X|\*))?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)\s*)((?:\s+|,\s*|\s*\|\|\s*)(=||!=|>|<|>=|=>|<=|=<|~|~>|\^)\s*(v?(0|[1-9]\d*|x|X|\*])(\.(0|[1-9]\d*|x|X|\*))?(\.(0|[1-9]\d*|x|X|\*]))?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)\s*)*$`
+	//+kubebuilder:Optional
+	// Version is an optional semver constraint on the package version. If not specified, the latest version available of the package will be installed.
+	// If specified, the specific version of the package will be installed so long as it is available in any of the content sources available.
+	// Examples: 1.2.3, 1.0.0-alpha, 1.0.0-rc.1
+	//
+	// For more information on semver, please see https://semver.org/
+	Version string `json:"version,omitempty"`
+
+	//+kubebuilder:validation:MaxLength:=48
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
+	// Channel constraint definition
+	Channel string `json:"channel,omitempty"`
+}
+
+type ExtensionSourceDirect struct {
+	// URL of the direct source
+	URL string `json:"url"`
```

Comment:
Can `ExtensionSourceDirect` be structured the same as `ImageSource` in Rukpak?

```suggestion
type ExtensionSourceDirect struct {
	Image *ImageSource `json:"image,omitempty"`
}

type ImageSource struct {
	// Ref contains the reference to a container image containing Bundle contents.
	Ref string `json:"ref"`
	// ImagePullSecretName contains the name of the image pull secret in the namespace that the provisioner is deployed.
	ImagePullSecretName string `json:"pullSecret,omitempty"`
}
```

Ideally it would be nice to have this same as the entire `BundleSource`: https://github.com/operator-framework/rukpak/blob/6a8a84c9aff05efaba7b05992704ad38462a7ee8/api/v1alpha2/bundle_types.go#L45. 

Just the URL seems ambiguous. Additionally, it would be nice to have consistency b/w what goes into Rukpak and the options that will be exposed here. With direct reference installations, we are anyway going to pass the values directly to BD API.

Reply:
Remember we're planning on using kapp-controller's App CR, not rukpak/BundleDeployment. But yes, we should see what fields to put in based on that.

Reply:
Right! Can we remove `ExtensionSourceDirect` from spec for now and add it later when we are ready for its implementation? We will have a better idea on what it should look like at that point in time.

Reply:
Makes sense

Reply:
I can remove `source.direct.url` from the spec... I'm assuming we still want to keep `source.package` as-is?

Reply:
👍 it contains all the catalog fields as present now, so looks good to me.